### DISCUSSION TITLE
Clickhouse kafka_max_block_size option

### DISCUSSION THREAD

**Discussion by knvk** [akvorado/akvorado#75] [Q&A]
Hi,
I'm doing some Alvarado performance/load tests with netflow exporter.
After collector hit 35k flows/sec, Kafka LAG grows up-to 8m messages. Read/write metrics are near 100k messages/sec.
```
# bin/kafka-consumer-groups.sh --bootstrap-server localhost:9092 --describe --group clickhouse | awk '{print $4 " "$5" "$6}' | grep -v "-"
CURRENT-OFFSET LOG-END-OFFSET  LAG
5234409406 5243845111 9435705
```
After some researches I found, that the reason is Clickhouse consumer is kind-of slow. 
Since I'm using singe exporter,  partition-balancing hash-function do **NOT** work (which is legal, it uses exporter-name), and producer sends all messages to the same partition, so increasing consumers number won't help. 

There are many similar issues in [Clickhouse repo](https://github.com/ClickHouse/ClickHouse/issues?q=lag+label%3Acomp-kafka)

For me increasing `kafka_max_block_size` became handy. 
Also, I ve increased:
```
<max_streams_to_max_threads_ratio>
<stream_flush_interval_ms>
```
Lag stoped at ~2M constantly, which is near 2 minutes in my environment. Read ratio up to 200k messages/sec.

options.
Notice, that `kafka_max_block_size` option don't affect already created tables, and have to be used on creation stage.
I did [fork](https://github.com/knvk/akvorado) with some changes if interested. 
[Perf tests](https://github.com/ClickHouse/ClickHouse/pull/11149#issuecomment-634582320) (not mine)

So, the question:
My goal is to achieve almost real-time processing. 
Is there any practies to speed-up Clickhouse in your project?

**Comment by vincentbernat** [comment]
It is my first time using ClickHouse, so I am not very skilled with it.

We could allow arbitrary options to be passed to Clickhouse Kafka engine if it helps (instead of enumarting all of them as options, I mean). Another possibility would be to not just use the exporter IP as a key for Kafka. I wonder what would be smart on this topic. It seemed natural to use the exporter IP, but I don't think this is helpful for anything. We could just use a random key. Maybe using the timestamp as a key would be another option which would help ClickHouse process stuff in batches (since it is sorting by TimeReceived, it is a bit less to sort).

You could easily try just using a random key by modifying the `Send()` method in `inlet/kafka/root.go`:

```go
// Send a message to Kafka.
func (c *Component) Send(exporter string, payload []byte) {
	c.metrics.bytesSent.WithLabelValues(exporter).Add(float64(len(payload)))
	c.metrics.messagesSent.WithLabelValues(exporter).Inc()
	key := make([]byte, 8)
	binary.BigEndian.PutUint64(key, rand.Uint64())
	c.kafkaProducer.Input() <- &sarama.ProducerMessage{
		Topic: c.kafkaTopic,
		Key:   sarama.ByteEncoder(key),
		Value: sarama.ByteEncoder(payload),
	}
}
```

**Comment by knvk** [comment]
It would be great to have arbitrary options for Clickhouse configuration, because this also makes possible to change even more table options. `PRIMARY KEY` and `PARTITION BY` for example. The bad thing is that some of these options takes effect at Creation time and some not.

I also think that point is to allow the user to change the producer's balancing algorithm via configuration is great. But should it be just per `exporter`/`random` or more ways, like `exporter+timestamp` tuple? I agree that exporter key is natural, if you have 8+ exporters. But if you have less, some of the consumers will be stalled. This moment should be more transparent for end-user.

If  possible, could you share what metrics are currently in your instance, and what flows/s and Kafka lag values ​​do you have?

Thanks

**Comment by vincentbernat** [comment]
On 2022-08-06 14:06, Konstantin Konov wrote:
> It would be great to have arbitrary options for Clickhouse 
> configuration, because this also makes possible to change even more 
> table options. |PRIMARY KEY| and |PARTITION BY| for example. The bad 
> thing is that some of these options takes effect at Creation time and 
> some not.

Too many flexibility comes with a maintenance price. If people really 
need more, they can just disable the ClickHouse component of the 
orchestrator (or not use the orchestrator at all).

> I also think that point is to allow the user to change the producer's 
> balancing algorithm via configuration is great. But should it be just 
> per |exporter|/|random| or more ways, like |exporter+timestamp| tuple? I 
> agree that exporter key is natural, if you have 8+ exporters. But if you 
> have less, some of the consumers will be stalled. This moment should be 
> more transparent for end-user.

Currently, I am leaning toward just using random. I have asked some 
people for insights on this.

> If possible, could you share what metrics are currently in your 
> instance, and what flows/s and Kafka lag values ​​do you have?

Check my latest presentation: 
https://vincent.bernat.ch/en/blog/2022-clickhouse-meetup-akvorado. There 
is a slide about this towards the end.

**Comment by vincentbernat** [comment]
Just pushed e5e63be58637 to use a 4-byte random value.
